### PR TITLE
Adding zipping and unzipping functionality to FileHostObject

### PR DESCRIPTION
**Pull request by Megala21**

### CODE REVIEW

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 1]
```diff
@@ -369,4 +369,178 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+     */
+    public static boolean jsFunction_unZip(Context cx, Scriptable thisObj, Object[] args, Function funObj)
+            throws ScriptException, IOException {
+        String functionName = "unZip";
+        int argsCount = args.length;
+        if (argsCount != 1) {
+            HostObjectUtil.invalidNumberOfArgs(hostObjectName, functionName, argsCount, false);
+        }
+        FileHostObject fho = (FileHostObject) thisObj;
+
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Remove these new lines.

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 2]
```diff
@@ -369,4 +369,178 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+     * @param cx      Context
+     * @param thisObj FileHostObject to be unzipped
+     * @param args    Path to unzip the zip file
+     * @param funObj  Function Object
+     * @throws ScriptException
+     */
+    public static boolean jsFunction_unZip(Context cx, Scriptable thisObj, Object[] args, Function funObj)
+            throws ScriptException, IOException {
+        String functionName = "unZip";
+        int argsCount = args.length;
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Don't we need to test args != null??

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 3]
```diff
@@ -369,4 +369,178 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+                        continue;
+                    }
+                    int hasParentDirs = name.lastIndexOf(File.separatorChar);
+                    dir = (hasParentDirs == -1) ? null : name.substring(0, hasParentDirs);
+                    if (dir != null) {
+                        mkdirs(outdir, dir);
+                    }
+                    BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(new File(outdir, name)));
+                    int count = -1;
+                    while ((count = zin.read(buffer)) != -1)
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Include the curly braces as a best practices for the while loop.

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 4]
```diff
@@ -369,4 +369,178 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+                        out.write(buffer, 0, count);
+                    out.close();
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Why this is closed twice? And close it in the finally block.

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java (hunk 2 above)
Comment (Megala21) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: I think when there are no arguments, it will come as empty array. So we do not need to check null. Similar way is followed in other methods in this class as well.

Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: ok.

File: components/jaggery-core/org.jaggeryjs.jaggery.core/src/main/java/org/jaggeryjs/jaggery/core/plugins/WebAppFileManager.java [hunk 5]
```diff
@@ -65,4 +65,22 @@ public File getFile(String path) throws ScriptException {
         }
         return file;
     }
+
+    @Override
+    public String getDirectoryPath(String path) throws ScriptException {
+        if (path.startsWith("file://")) {
```
Comment (sinthuja) [inline components/jaggery-core/org.jaggeryjs.jaggery.core/src/main/java/org/jaggeryjs/jaggery/core/plugins/WebAppFileManager.java]: Include the file:// as constant.

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 6]
```diff
@@ -369,4 +369,196 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+            } catch (IOException ex) {
+                log.error("Cannot unzip the file " + ex);
+                throw new IOException(ex);
+            } finally {
+                if (zin != null) {
+                    zin.close();
+                }
+            }
+        } else {
+            log.error("Zip File not exists");
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: 'File' should start with simple f

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 7]
```diff
@@ -369,4 +369,196 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+                        mkdirs(outdir, name);
+                        continue;
+                    }
+                    int hasParentDirs = name.lastIndexOf(File.separatorChar);
+                    dir = (hasParentDirs == -1) ? null : name.substring(0, hasParentDirs);
+                    if (dir != null) {
+                        mkdirs(outdir, dir);
+                    }
+                    BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(new File(outdir, name)));
+                    try {
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Not necessarily you need to have two try blocks, this can be done with only single try block as the inner try block doesn't do anything additional.

File: components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java [hunk 8]
```diff
@@ -369,4 +369,196 @@ public static Object jsFunction_listFiles(Context cx, Scriptable thisObj, Object
...
+    /**
+     * To add a folder to zip
+     *
+     * @param path      Path of the file or folder from root directory of zip
+     * @param srcFolder Source folder to be made as zip
+     * @param zip       ZipOutputStream
+     */
+    private static void addFolderToZip(String path, String srcFolder, ZipOutputStream zip) throws IOException {
+        File folder = new File(srcFolder);
+        if (path.equals("")) {
```
Comment (sinthuja) [inline components/hostobjects/org.jaggeryjs.hostobjects.file/src/main/java/org/jaggeryjs/hostobjects/file/FileHostObject.java]: Can simply use isEmpty.
